source/opt/const_folding_rules.cpp

```diff
@@ -120,6 +120,84 @@ ConstantFoldingRule FoldExtractWithConstants() {
   };
 }
 
+// Folds an OpcompositeInsert where input is a composite constant.
+ConstantFoldingRule FoldInsertWithConstants() {
+  return [](IRContext* context, Instruction* inst,
+            const std::vector<const analysis::Constant*>& constants)
+             -> const analysis::Constant* {
+    analysis::ConstantManager* const_mgr = context->get_constant_mgr();
+    const analysis::Constant* object = constants[0];
+    const analysis::Constant* composite = constants[1];
+    if (object == nullptr || composite == nullptr) {
+      return nullptr;
+    }
+
+    // If there is more than 1 index, then each additional constant used by the
+    // index will need to be recreated to use the inserted object
+    std::vector<const analysis::Constant*> chain;
+    std::vector<const analysis::Constant*> components;
+    const analysis::Type* type;
+
+    // work down hierarchy and add all the indexes, not including the final
+    // index
+    for (uint32_t i = 2; i < inst->NumInOperands(); ++i) {
+      if (i != inst->NumInOperands() - 1) {
+        chain.push_back(composite);
+      }
+      const uint32_t index = inst->GetSingleWordInOperand(i);
+      components = composite->AsCompositeConstant()->GetComponents();
+      type = composite->AsCompositeConstant()->type();
+      composite = components[index];
+    }
+
+    // final index in hierarchy is inserted with new object
+    const uint32_t final_index =
+        inst->GetSingleWordInOperand(inst->NumInOperands() - 1);
+    std::vector<uint32_t> ids;
+    for (size_t i = 0; i < components.size(); i++) {
+      const analysis::Constant* constant =
+          (i == final_index) ? object : components[i];
+      Instruction* member_inst = const_mgr->GetDefiningInstruction(constant);
+      ids.push_back(member_inst->result_id());
+    }
+    const analysis::Constant* new_constant = const_mgr->GetConstant(type, ids);
+
+    // work backwards up the chain and replace each index with new constant
+    for (size_t i = chain.size(); i > 0; i--) {
+      // Need to insert any previous instruction into the module first
+      // TODO - Is there a better way to find this position
+      Module::inst_iterator* pos = nullptr;
```

Comment:
You could also insert at the top of the constants declared in the module, I think.

Reply:
Ya, I guess I was going for keeping the instructions next to each other, but I also like the idea of making the code simpler and just putting at the top (since I should have an easy reference to its iterator)

Reply:
So I tried this approach and the issue currently there is only an iterator to the `Types` which ends up inserting it above where the types are declared. Updated comment and think this is the best way currently to find the instruction

Reply:
friendly ping

Reply:
Sorry!  LGTM.